edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/brokerConnection/MqttBrokerConnector.cs

```diff
@@ -115,7 +115,19 @@ public async Task DisconnectAsync()
                             c.MqttMsgPublishReceived -= this.ForwardPublish;
                             c.ConnectionClosed -= this.TriggerReconnect;
 
-                            c.Disconnect();
+                            if (c.IsConnected)
```

Comment:
Can this check be moved to Disconnect method?

Reply:
you mean the c.IsConnected check? This class maintains its own state (connected/disconnected), and when we get here, this state is "connected". What may happen is that the broker is already down, so the underlying socket is already disconnected (even if this class is still in "connected" state). This check is on the mqtt library I use, that may already noticed that  it does not have connection (and down in the code there is a TryReconnect() method, that may be running and it tries to re-connect). The object disposed exception came when the c.Disconnect() call was made in this state, that is why the check before the call. Also, there might be a race condition, that the c.IsConnected() still believes that the library is connected, but by the time the c.Disconnect() call gets there to close the socket, the socket is cleaned up (because the broker dropped the connection)